src/floki_mochi_html.erl

```diff
@@ -215,38 +226,44 @@ tree_data([{data, Data, Whitespace} | Rest], AllWhitespace, Acc) ->
 tree_data(Rest, AllWhitespace, Acc) ->
     {iolist_to_binary(lists:reverse(Acc)), AllWhitespace, Rest}.
 
-tree([], Stack) ->
+tree([], Stack, _Opts) ->
     {destack(Stack), []};
-tree([{end_tag, Tag} | Rest], Stack) ->
-    case destack(norm(Tag), Stack) of
+tree([{end_tag, Tag} | Rest], Stack, Opts) ->
+    case destack(norm(Tag, Opts), Stack) of
         S when is_list(S) ->
-            tree(Rest, S);
+            tree(Rest, S, Opts);
         Result ->
             {Result, []}
     end;
-tree([{start_tag, Tag, Attrs, true} | Rest], S) ->
-    tree(Rest, append_stack_child(norm({Tag, Attrs}), S));
-tree([{start_tag, Tag, Attrs, false} | Rest], S) ->
-    tree(Rest, stack(norm({Tag, Attrs}), S));
-tree([T = {pi, _Tag, _Attrs} | Rest], S) ->
-    tree(Rest, append_stack_child(T, S));
-tree([T = {comment, _Comment} | Rest], S) ->
-    tree(Rest, append_stack_child(T, S));
-tree(L = [{data, _Data, _Whitespace} | _], S) ->
+tree([{start_tag, Tag, Attrs, true} | Rest], S, Opts) ->
+    tree(Rest, append_stack_child(norm({Tag, Attrs}, Opts), S), Opts);
+tree([{start_tag, Tag, Attrs, false} | Rest], S, Opts) ->
+    tree(Rest, stack(norm({Tag, Attrs}, Opts), S), Opts);
+tree([T = {pi, _Tag, _Attrs} | Rest], S, Opts) ->
+    tree(Rest, append_stack_child(T, S), Opts);
+tree([T = {comment, _Comment} | Rest], S, Opts) ->
+    tree(Rest, append_stack_child(T, S), Opts);
+tree(L = [{data, _Data, _Whitespace} | _], S, Opts) ->
     case tree_data(L, true, []) of
         {_, true, Rest} ->
-            tree(Rest, S);
+            tree(Rest, S, Opts);
         {Data, false, Rest} ->
-            tree(Rest, append_stack_child(Data, S))
+            tree(Rest, append_stack_child(Data, S), Opts)
     end;
-tree([{doctype, _} | Rest], Stack) ->
-    tree(Rest, Stack).
-
-norm({Tag, Attrs}) ->
-    {norm(Tag), [{norm(K), iolist_to_binary(V)} || {K, V} <- Attrs], []};
-norm(Tag) when is_binary(Tag) ->
+tree([{doctype, _} | Rest], Stack, Opts) ->
+    tree(Rest, Stack, Opts).
+
+norm({Tag, Attrs}, Opts) ->
+    Attrs = [{norm(K, Opts), iolist_to_binary(V)} || {K, V} <- Attrs],
+    case lists:keyfind(attributes_as_maps, 1, Opts) of
+        {attributes_as_maps, true} ->
+            {norm(Tag, Opts), maps:from_list(Attrs), []};
```

Comment:
Documentation on `maps:from_list` says

> If the same key appears more than once, the latter (right-most) value is used and the previous values are ignored.

This means for duplicate attributes the _last_ attribute is used in the map.

The HTML spec suggests that it's a parse error when you encounter multiple identical attributes ([duplicate-attribute parse error](https://html.spec.whatwg.org/#parse-error-duplicate-attribute)), and that the parser ignores the duplicates (keeping the first).

Reply:
Great catch! Fixed in 708ad4406012097d1514dee328f1f08ad4ccefe3
Thank you! :purple_heart: